# Computer Graphik: Projektive Geometrie III - Übung

I.BA_CG, SW 05

Léonard Wyrsch

## Aufgabe 1: Rotation im dreidimensionalen Raum
Der Würfel mit der Seitenlänge a = 1 hat eine Ecke im Nullpunkt und befindet sich vollständig im ersten Oktanten ({(x, y, z) ∈ R<sup>3</sup> | x ≥ 0 ∧ y ≥ 0 ∧ z ≥ 0}). Wie lautet die (homogene) Transformationsmatrix, welche diesen Würfel im mathematisch positiven Sinn um 45° um die y-Achse dreht. Wie lauten dann die neuen Koordinaten der acht Eckpunkte?

        0 1 0 1 0 1 0 1         cos(φ) 0 sin(φ) 0
    A = 0 0 1 1 0 0 1 1     R =     0  1     0  0   φ = 45°
        0 0 0 0 1 1 1 1        -sin(φ) 0 cos(φ) 0
        1 1 1 1 1 1 1 1             0  0     0  1

                         0   1   0   1    0    1  0    1
                     A = 0   0   1   1    0    0  1    1
                         0   0   0   0    1    1  1    1
                         1   1   1   1    1    1  1    1

        1/√2 0 1/√2 0    0  1/√2 0  1/√2 1/√2 √2 1/√2 √2
    R =  0   1  0   0    0   0   1   1    0    0  1    1 = A'
       -1/√2 0 1/√2 0    0 -1/√2 0 -1/√2 1/√2  0 1/√2  0
         0   0  0   1    1   1   1   1    1    1  1    1

## Aufgabe 2: Rotation im dreidimensionalen Raum
Die Gerade g geht durch die Punkte P(2,0,0) und Q(0,4,0). Sie liegt also in der x-y-Ebene. Sei δ die Drehung um die Gerade g mit dem Drehwinkel φ = 55°. Gesucht ist die Matrix D dieser Drehung

**Hinweis:** Man erhält die gesuchte Matrix, wenn man ρ wie folgt zusammensetzt:
- Zuerst eine Translation τ, welche g in eine Gerade g<sup>\*</sup> durch den Nullpunkt schiebt
- Nach einer geeigneten Rotation R um die z-Achse kommt g<sup>\*</sup> auf die x-Achse zu liegen
- Rotation η um die x-Achse mit dem Winkel φ.
Dann ist δ = τ<sup>−1</sup> ◦ ρ<sup>−1</sup> ◦ η ◦ ρ ◦ τ.

    P(2, 0, 0)      Q(0, 4, 0)      φ = 55°

        1 0 0 -2           1 0 0 2
    τ = 0 1 0  0    τ^-1 = 0 1 0 0
        0 0 1  0           0 0 1 0
        0 0 0  1           0 0 0 1
    
    P'(0, 0, 0)     Q'(-2, 4, 0)
    m' = 4/-2 = -2 = tan(φ') => sin(φ') = 2/√5, cos(φ') = 1/√5

        1/√5 -2/√5 0 0              1/√5 2/√5 0 0
    ρ = 2/√5  1/√5 0 0      ρ^-1 = -2/√5 1/√5 0 0
         0     0   1 0               0    0   1 0
         0     0   0 1               0    0   0 1

        1      0         0   0
    η = 0 cos(55°) -sin(55°) 0
        0 sin(55°)  cos(55°) 0
        0      0         0   1

    δ = τ^−1 ◦ ρ^−1 ◦ η ◦ ρ ◦ τ

In [2]:
import math
import numpy as np

T = np.array([[1, 0, 0, -2],
              [0, 1, 0,  0],
              [0, 0, 1,  0],
              [0, 0, 0,  1]])

T_inv = np.array([[1, 0, 0, 2],
                  [0, 1, 0, 0],
                  [0, 0, 1, 0],
                  [0, 0, 0, 1]])

R = np.array([[1/math.sqrt(5), -2/math.sqrt(5), 0, 0],
              [2/math.sqrt(5),  1/math.sqrt(5), 0, 0],
              [             0,               0, 1, 0],
              [             0,               0, 0, 1]])

R_inv = R.transpose()

rad = math.radians(55)
N = np.array([[1,             0,              0, 0],
              [0, math.cos(rad), -math.sin(rad), 0],
              [0, math.sin(rad),  math.cos(rad), 0],
              [0,             0,              0, 1]])

result = np.identity(4)
for mat in [T, R, N, R_inv, T_inv]:
    result = np.matmul(result, mat)

print(result)

[[ 0.65886115  0.17056943  0.73267186 -0.6822777 ]
 [ 0.17056943  0.91471529 -0.36633593  0.34113885]
 [-0.73267186  0.36633593  0.57357644 -1.46534372]
 [ 0.          0.          0.          1.        ]]


        0.65886115 0.17056943  0.73267186 -0.6822777
    δ = 0.17056943 0.91471529 -0.36633593  0.34113885
       -0.73267186 0.36633593  0.57357644 -1.46534372
        0.         0.          0.          1.

## Aufgabe 3: Transformation im dreidimensionalen Raum
Die 3D-Transformation σ ist gegeben durch die Matrix

              1 4 -8 -28
    M = 1/9 * 4 7  4  14
             -8 4  1 -28
              0 0    0 9

- (a) Berechnen Sie die Bildpunkte A<sup>\*</sup> und B<sup>\*</sup> von A(2|3|-5) und B(6.2|-4.8|-1.7) und bestätigen Sie, dass der Abstand erhalten bleibt, d.h. A<sup>\*</sup>B<sup>\*</sup> = AB.
- (b) Es stellt sich heraus, dass σ eine Spiegelung ist. Bestimmen Sie die Gleichung der Ebene ε, an welcher gespiegelt wird.

**Hinweis:** Die Punkte in ε sind dadurch gekennzeichnet, dass sie bei der Transformation fest bleiben.

a)

                               2        6.2
                           A = 3   B = -4.8
                              -5       -1.7
                               1        1

        1/9 4/9 -8/9 -28/9   26/9     -27.4/9
    M = 4/9 7/9  4/9  14/9   23/9      -1.6/9
       -8/9 4/9  1/9 -28/9  -37/9     -98.5/9
         0   0    0     1      1        1

    A'(26/9 | 23/9 | -37/9)
    B'(-27.4/9 | -1.6/9 | -98.5/9)

         2 - 6.2   -4.2
    AB = 3 + 4.8 =  7.8             |AB| = √89.37
        -5 + 1.7   -3.3

           26/9 + 27.4/9   53.4/9
    A'B' = 23/9 +  1.6/9 = 24.6/9   |A'B'| = √(7238.97/81) = √89.37 = |AB|
          -37/9 + 98.5/9   61.5/9

---
b)

     M * [x y z 1]T = [x y z 1]T => (M-I)*[x y z 1]T = 0

      1/9 4/9 -8/9 -28/9   1 0 0 0  -8/9  4/9 -8/9 -28/9
      4/9 7/9  4/9  14/9 - 0 1 0 0 = 4/9 -2/9  4/9  14/9
     -8/9 4/9  1/9 -28/9   0 0 1 0  -8/9  4/9 -8/9 -28/9
       0   0    0     1    0 0 0 1    0    0    0     0

    -8/9  4/9 -8/9 -28/9 0           -8/9 4/9 -8/9 -28/9 0  *-9
     4/9 -2/9  4/9  14/9 0  +1/2[1]    0   0    0     0  0
    -8/9  4/9 -8/9 -28/9 0  -1[1]      0   0    0     0  0
      0    0    0     0  0             0   0    0     0  0

    8 -4 8 28 0  /4  2 -1 2 7 0
    0  0 0  0 0      0  0 0 0 0
    0  0 0  0 0      0  0 0 0 0
    0  0 0  0 0      0  0 0 0 0

    ε: 2x - y + 2z + 7 = 0

## Aufgabe 4: Transformation im dreidimensionalen Raum
Die Punkte A(0,0,0), B(6,6,3), C(0,9,9), D(-6,3,6), E(3,-6,6), F(9,0,9), G(3,3,15) und H(-3,-3,12) sind die Ecken eines Würfels. Sei π die Parallel-Projektion auf die x-z-Ebene entlang des Vektors(1,1,0).

Gesucht
- (a) die Matrix der Projektion und
- (b) das Abbild des Würfels in der x-z-Ebene.

## Aufgabe 5: Transformation im dreidimensionalen Raum
Die Punkte A(0,0,0), B(6,6,3), C(0,9,9), D(-6,3,6), E(3,-6,6), F(9,0,9), G(3,3,15) und H(-3,-3,12) sind die Ecken eines Würfels. Sei π die perspektivische Projektion auf die x-y-Ebene mit Zentrum Z(2,4,-3).

Gesucht sind
- (a) die Matrix der Projektion und
- (b) das Abbild des Würfels in der x-y-Ebene.

## Aufgabe 6: Perspektivische Projektion
Wir betrachten die perspektivische Projektion auf die Bildebene ε : z = -5 mit dem Zentrum im Nullpunkt. Das Sichtvolumen wird so festgelegt: Nach vorne durch ε : z = -5 und nach hinten durch ε : z = -10. Seitlich durch das rechteckige Fenster mit den Ecken P(-3,-2,-5) und Q(3,3,-5).

- (a) Bestimmen Sie die Matrix T, welche die Transformation auf das kanonische Sichtvolumen erzeugt.
- (b) Berechnen Sie die Bildpunkte von A(-4.5,4.2,-8) und B(1.6,-2.4,-5.5) und entscheiden Sie dann, ob A und B innerhalb oder ausserhalb des Sichtvolumens liegen.